Source/AwsCommonRuntimeKit/mqtt/Mqtt5Client.swift

```diff
@@ -121,44 +121,154 @@ public typealias OnWebSocketHandshakeInterceptComplete = (HTTPRequestBase, Int32
     init(clientOptions options: MqttClientOptions) throws {
 
         try options.validateConversionToNative()
 
-        self.callbackCore = MqttCallbackCore(
-            onPublishReceivedCallback: options.onPublishReceivedFn,
-            onLifecycleEventStoppedCallback: options.onLifecycleEventStoppedFn,
-            onLifecycleEventAttemptingConnect: options.onLifecycleEventAttemptingConnectFn,
-            onLifecycleEventConnectionSuccess: options.onLifecycleEventConnectionSuccessFn,
-            onLifecycleEventConnectionFailure: options.onLifecycleEventConnectionFailureFn,
-            onLifecycleEventDisconnection: options.onLifecycleEventDisconnectionFn,
-            onWebsocketInterceptor: options.onWebsocketTransform)
+        self.onPublishReceivedCallback = options.onPublishReceivedFn ?? { (_) in return }
+        self.onLifecycleEventStoppedCallback = options.onLifecycleEventStoppedFn ?? { (_) in return}
+        self.onLifecycleEventAttemptingConnect = options.onLifecycleEventAttemptingConnectFn ?? { (_) in return}
+        self.onLifecycleEventConnectionSuccess = options.onLifecycleEventConnectionSuccessFn ?? { (_) in return}
+        self.onLifecycleEventConnectionFailure = options.onLifecycleEventConnectionFailureFn ?? { (_) in return}
+        self.onLifecycleEventDisconnection = options.onLifecycleEventDisconnectionFn ?? { (_) in return}
+        self.onWebsocketInterceptor = options.onWebsocketTransform
 
         guard let rawValue = (options.withCPointer(
-            userData: self.callbackCore.callbackUserData()) { optionsPointer in
+            userData: Unmanaged<Mqtt5ClientCore>.passRetained(self).toOpaque()) { optionsPointer in
                 return aws_mqtt5_client_new(allocator.rawValue, optionsPointer)
             }) else {
             // failed to create client, release the callback core
-            self.callbackCore.release()
+            Unmanaged<Mqtt5ClientCore>.passUnretained(self).release()
             throw CommonRunTimeError.crtError(.makeFromLastError())
         }
         self.rawValue = rawValue
     }
 
-    deinit {
-        self.callbackCore.close()
-        aws_mqtt5_client_release(rawValue)
-    }
-
+    /// Notifies the Mqtt5Client that you want it maintain connectivity to the configured endpoint.
+    /// The client will attempt to stay connected using the properties of the reconnect-related parameters
+    /// in the Mqtt5Client configuration on client creation.
+    ///
+    /// - Throws: CommonRuntimeError.crtError
     public func start() throws {
-        if rawValue != nil {
+        try self.rwlock.read {
+            // Validate close() has not been called on client.
+            guard let rawValue = self.rawValue else {
+                // TODO add new error type for client closed
+                throw CommonRunTimeError.crtError(CRTError.makeFromLastError())
+            }
```

Comment:
We can pull this into a helper closure, since every function will need to do this.

Reply:
The `rawValue` null validation would be removed in a later [PR](https://github.com/awslabs/aws-crt-swift/pull/264/files#diff-56edcc58b3d774bc93486d6f0f62fd6ddb0bac7bdcb18b268cdbd02bc6e54532L182). The final version would be just one line code:
```
 try self.rwlock.read {

}
```
I will keep it as it for now.